### ISSUE TITLE
Unable to load tokenizer.json

### ISSUE THREAD

**Issue by ajoshibytesarena** [georg-jung/FastBertTokenizer#92]
for the model I am using currently, I do not get vocab.txt, instead I only have tokenizer.json file.

Currently I have finetuned version of model: sentence-transformers/paraphrase-xlm-r-multilingual-v1 (After finetunning, I only have model files as tokenizer.json, tokenizer_config.json, special_tokens_map.json, model.safetensors and config.json)

in .net code i am trying to Initialize tokenizer using code
Approach 1:

var tok = new BertTokenizer();
await tok.LoadTokenizerJsonAsync(
            @"E:\Project_GTI_CI\onnx_model_with_dotnet_31jan\bert_model_files\tokenizer\tokenizer.json"
        );
        

but getting error as,

Unhandled exception. System.Text.Json.JsonException: The JSON value could not be converted to System.Collections.Generic.Dictionary`2[System.String,System.Int32]. Path: $.model.vocab | LineNumber: 177 | BytePositionInLine: 14.
   at System.Text.Json.ThrowHelper.ThrowJsonException_DeserializeUnableToConvertValue(Type propertyType)
   at System.Text.Json.Serialization.JsonDictionaryConverter`3.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, TDictionary& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
   at System.Text.Json.Serialization.JsonConverter`1.TryReadAsObject(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, Object& value)
   at System.Text.Json.Serialization.Converters.LargeObjectWithParameterizedConstructorConverter`1.ReadAndCacheConstructorArgument(ReadStack& state, Utf8JsonReader& reader, JsonParameterInfo jsonParameterInfo)
   at System.Text.Json.Serialization.Converters.ObjectWithParameterizedConstructorConverter`1.ReadConstructorArgumentsWithContinuation(ReadStack& state, Utf8JsonReader& reader, JsonSerializerOptions options)
   at System.Text.Json.Serialization.Converters.ObjectWithParameterizedConstructorConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
   at System.Text.Json.Serialization.JsonConverter`1.TryReadAsObject(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, Object& value)
   at System.Text.Json.Serialization.Converters.LargeObjectWithParameterizedConstructorConverter`1.ReadAndCacheConstructorArgument(ReadStack& state, Utf8JsonReader& reader, JsonParameterInfo jsonParameterInfo)
   at System.Text.Json.Serialization.Converters.ObjectWithParameterizedConstructorConverter`1.ReadConstructorArgumentsWithContinuation(ReadStack& state, Utf8JsonReader& reader, JsonSerializerOptions options)
   at System.Text.Json.Serialization.Converters.ObjectWithParameterizedConstructorConverter`1.OnTryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value)
   at System.Text.Json.Serialization.JsonConverter`1.TryRead(Utf8JsonReader& reader, Type typeToConvert, JsonSerializerOptions options, ReadStack& state, T& value, Boolean& isPopulatedValue)
   at System.Text.Json.Serialization.JsonConverter`1.ReadCore(Utf8JsonReader& reader, T& value, JsonSerializerOptions options, ReadStack& state)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.ContinueDeserialize(ReadBufferState& bufferState, JsonReaderState& jsonReaderState, ReadStack& readStack, T& value)
   at System.Text.Json.Serialization.Metadata.JsonTypeInfo`1.DeserializeAsync(Stream utf8Json, CancellationToken cancellationToken)
   at FastBertTokenizer.BertTokenizer.LoadTokenizerJsonAsync(Stream tokenizerJsonStream)
   at FastBertTokenizer.BertTokenizer.LoadTokenizerJsonAsync(String tokenizerJsonFilePath)
   at Program.Main(String[] args) in E:\Project_GTI_CI\onnx_model_with_dotnet_31jan\MLONNX\Program.cs:line 295
   at Program.<Main>(String[] args)

Approach 2: I have prepared vocab.text file programmatically using python as
import json

# Load JSON file
with open(r"tokenizer.json", "r", encoding="utf-8") as file:
    data = json.load(file)
# Extract keys from "vocab"
vocab = []
vocab_list = data["model"]["vocab"]
for voc in vocab_list:
    vocab.append(voc[0])
print(len(vocab))

with open("vocab.txt", "w", encoding="utf-8") as output_file:
    output_file.write("\n".join(vocab))

and then used it in .net code and initialize tokenizer as 
var tok = new BertTokenizer();

await tok.LoadVocabularyAsync(
    vocabTxtFilePath: @"E:\Project_GTI_CI\onnx_model_with_dotnet_31jan\bert_model_files\tokenizer\vocab.txt",
    convertInputToLowercase: true,
    unknownToken: "<unk>",
    clsToken: "<s>",
    sepToken: "</s>",
    padToken: "<pad>"
);

this code worked but giving wrong Input IDs and Attention Mask values than expected. What is the proper way to use tokenizer.json file ?

**Comment by georg-jung** [comment]
Your approach seems fine in general. Can you share your tokenizer.json file? The issue seems to be about the content/format of that file. To verify this, you could e.g. try to load [this tokenizer.json file](https://huggingface.co/google-bert/bert-base-uncased/blob/main/tokenizer.json). To go further you could compare your file's structure with the one I linked.

I'd be happy to support more tokenizer.json variants if this strictness on FastBertTokenizer's side is not required.

**Comment by ajoshibytesarena** [comment]
this is the format of tokenizer file that I am using,

[tokenizer (3).json](https://github.com/user-attachments/files/18618361/tokenizer.3.json)

**Comment by georg-jung** [comment]
This tokenizer config is not about a BERT tokenizer. It contains e.g.

```
"model":{"type":"Unigram"
```

For bert tokenization, which FastBertTokenizer is built for, this would read 

```
"model":{"type":"WordPiece"
```

FastBertTokenizer contains a code path that is meant to provide you with a nicer error message in this case:

https://github.com/georg-jung/FastBertTokenizer/blob/4d39d6c2d88da3a3978176006d14c43c41f30c13/src/FastBertTokenizer/BertTokenizer.LoadTokenizerJson.cs#L76

But, given the tokenizer config & its JSON representation is different in structure, deserialization already fails and this code path can not be reached. I could maybe add a nicer exception wrapping this kind of deserialization exceptions to make this more clear.

I'm not sure if a tokenizer for your tokenizer.json currently exists in the .NET world. At least [Microsoft.ML.Tokenizers does not seem to support](https://github.com/dotnet/machinelearning/issues/7186) unigram as of now - but [adding support is considered](https://github.com/dotnet/machinelearning/issues/7294#issuecomment-2481609555).

**Comment by georg-jung** [comment]
One option you have is to wrap huggingface's rust tokenizer lib. This would however require some considerable effort. You can take a look at https://github.com/georg-jung/FastBertTokenizer/tree/master/src/HuggingfaceTokenizer/RustLibWrapper to get started.
